### PR TITLE
Fix bugs of pkey_dh_copy()

### PR DESCRIPTION
**Pull request by jakkdu**
This is follow up PR for #647.

### CODE REVIEW

**Comment by jakkdu** [comment]
Ping to @mattcaswell @ajmohan @richsalz.
Sorry I deleted my forked repo.
So I made a follow up PR for #647.

File: crypto/dh/dh_pmeth.c [hunk 1]
```diff
@@ -112,6 +112,17 @@ static int pkey_dh_init(EVP_PKEY_CTX *ctx)
     return 1;
 }
 
+static void pkey_dh_cleanup(EVP_PKEY_CTX *ctx)
+{
+    DH_PKEY_CTX *dctx = ctx->data;
+    if (dctx) {
```
Comment (mattcaswell) [inline crypto/dh/dh_pmeth.c]: Sorry. Didn't spot this before. Please make this `dctx != NULL`

**Comment by mattcaswell** [comment]
I am happy with this aside from the one minor comment

**Comment by jakkdu** [comment]
@mattcaswell applied

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
We need another team member review now.

**Comment by kroeckx** [comment]
I have a problem with this. All the copy functions are either not documented, or just says that it copies things.  It clearly does not document that dst can be reused and what the state of it is on failure.

In any case, I think it's a bad idea to free() things without setting the pointer to NULL for an object that's probably not in a consistent state.

**Comment by jakkdu** [comment]
@kroeckx then do you want to set NULL in all freed objects in pkey_dh_cleanup()?
If it is, then I think it is not a good idea. It can produce unnecessary overhead. 

A developer who uses pkey_dh_cleanup() should check the return value,
if it returns 0, which means an error, then pkey_dh_cleanup() guarantees that copying object is failed and no leaked memory, but not the state of an object.

**Comment by kroeckx** [comment]
I'm not sure I understand what you're saying.  I assume that the normal way to use the copy function is that you first allocate and initialize dst, then try try to copy it, and if it failed try to free it.

I assume that initializing dst sets the pointers inside the struct to NULL.  I assume that free()ing dst will try to free the pointers inside the struct.  Since they're already free()d you'd end up with a double free if you don't set them to NULL.

If those assumptions are wrong, it clearly needs to be documented.

There are also at least _documented_ cases where dst does not need to be a just initialized object but can be an existing object of which you want to replace some parameters.  In that case I expect that _at least_ dst ends up in a consistent state (basic exception safety) but would prefer that the behaviour of the function is that it either succeeds or does not changed dst at all (strong exception safety).

**Comment by jakkdu** [comment]
@kroeckx I see. I think I understand what you mean. 
I think we should not free internal data because normal way to use copy function is
init -> copy -> if failed -> free.

Therefore if we free it internally, then it may cause double free, right? 
I think we do not need to free it like pkey_rsa_copy()
So, if an error occurs just return 0 without free.

**Comment by jakkdu** [comment]
@kroeckx I committed another one. I think it is what you mention, right?

**Comment by jakkdu** [comment]
Then I think we don't need to move pkey_dh_cleanup(). 
But I think it is not an issue.

**Comment by kroeckx** [comment]
+1

**Comment by FdaSilvaYY** [comment]
Tips : run a "git rebase -i HEAD~4 " to squash these four commits...

**Comment by jakkdu** [comment]
@FdaSilvaYY +1

**Comment by mattcaswell** [comment]
Merged. Thanks.

**Pull request closed without merging**